File: Algorithm/Python/144. Binary Tree Preorder Traversal.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-06-01 22:24:37
# @Last modified by:   WuLC
# @Last Modified time: 2016-06-01 22:34:46
# @Email: liangchaowu5@gmail.com


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


# method 1,recursively
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []
        if root == None:
            return result
        result.append(root.val)
        result += self.preorderTraversal(root.left)
        result += self.preorderTraversal(root.right)
        return result
            
# method 2,iteratively
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        stack, result = [], []
        curr_node = root
        while len(stack) != 0 or curr_node!= None:
            if curr_node != None:
                result.append(curr_node.val)
                stack.append(curr_node)
                curr_node = curr_node.left
            else:
                tmp = stack.pop()
                curr_node = tmp.right
        return result       

File: Algorithm/Java/144. Binary Tree Preorder Traversal.java
/*
* @Author: WuLC
* @Date:   2017-03-01 08:54:19
* @Last Modified by:   WuLC
* @Last Modified time: 2017-03-01 09:08:56
* @Email: liangchaowu5@gmail.com
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
 
 // recusrive 
public class Solution 
{
    public List<Integer> preorderTraversal(TreeNode root) 
    {
        List<Integer> curr = new ArrayList<Integer>();
        if (root == null) return curr;
        curr.add(root.val);
        List<Integer> left = preorderTraversal(root.left);
        List<Integer> right = preorderTraversal(root.right);
        curr.addAll(left);
        curr.addAll(right);
        return curr;
    }
}

// use stack, non- recursive
public class Solution 
{
    public List<Integer> preorderTraversal(TreeNode root) 
    {
        List<Integer> result = new ArrayList<Integer>();
        Stack<TreeNode> nodes = new Stack<TreeNode>();
        TreeNode curr = root;
        while(nodes.size()>0 || curr!=null)
        {
            if (curr != null) 
            {
                result.add(curr.val);
                nodes.push(curr);
                curr = curr.left;
            }
            else
            {
                curr = nodes.pop();
                curr = curr.right;
            }
        }
        return result;
    }
}


